### PR TITLE
an arguable improvement to monix usage

### PR DESCRIPTION
**Pull request by SwiftEngineer**
Changelist:
- Wrap main with TaskApp
- Instantiate scheduler on the fly using `cats.effect.Resource` cause we like closin' stuff
- Try out reactivex-based (using monix Observable) approach for ParallelismDemo
- Add [jctools-graal-native](https://github.com/SwiftEngineer/jctools-graal-native) to fix segmentation faults that would otherwise occur due to monix's indirect use of `sun.misc.Unsafe` (and friends).
- update scalatest to `3.2.0`

### CODE REVIEW

**Comment by SwiftEngineer** [comment]
I'm gonna try to annotate this PR's changes and provide some benchmark times tomorrow, but thought I'd push this up as an early preview anyway

**Comment by SwiftEngineer** [comment]
Benchmark times (m:ss):

All benchmarks were run the same as in my last PR. It still includes container spin up time. The only difference this time was I only *provided each container with 5 cpus*. This was because I wanted to make sure that the effect of setting parallelism higher than the amount of available cpu matched my expectations. Spoiler alert: I had my hopes a little high.

### Run times using previous code (master directly after merging the update to 20.2.0 of Graal)

| Parallelism |  Iterations  | Run 1 |  Run 2 | Run 3 |
| --------------- | -------------- | --------------- | ---------------- | --------------- |
| 1 | 10,000,000 | 1:10.49 | 1:09.22 | 0:59.23 |
| 2 | 10,000,000 | 0:52.46 | 0:55.03 | 0:57.58 |
| 5 | 10,000,000 | 1:01.10 | 0:59.94 | 1:01.86 |
| 10 | 10,000,000 | 1:07.49 | 1:03.49 | 1:08.83 | 

### Run times with new code

| Parallelism |  Iterations  | Run 1 |  Run 2 | Run 3 |
| --------------- | -------------- | --------------- | ---------------- | --------------- |
| 1 | 10,000,000 | 0:02.89 | 0:02.82 | 0:02.88 |
| 2 | 10,000,000 | 0:27.80 | 0:27.57 | 0:26.64 |
| 5 | 10,000,000 | 0:16.99 | 0:17.00 | 0:17.15 |
| 10 | 10,000,000 | 0:21.95 | 0:21.63 | 0:21.71 |  

### Analysis

At first, one might scratch their head a bit at these numbers. Here's some answers to some burning questions you may or may not have:

*Q*: Why is there such a huge difference in performance?
*A*: `Task.parSequence` wasn't designed for task counts as high as what we're using above. It appears as though the majority of the old code's run time is dedicated to composing the Task instance that is returned by the `ParallelismDemo.approximatePi` function, since the 30 second timeout that we place on the resulting Future was never triggered. Meaning the evaluation of the `Task` clearly didn't start until pretty late in the run time.

*Q*: Why is it that setting the parallelism to 1 perform so much better than any other setting?
*A*: Monix is kinda smart. When parallelism is set to 1, it is able to use a much more [performant subscriber](https://github.com/monix/monix/blob/series/3.x/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapTaskObservable.scala#L49-L63).

*Q*: Why is parallelism 5 faster than parallelism 10?
*A*: At the top of this benchmark, I mentioned we ran with 5 cpus. so of course, setting parallelism higher than 5 wasn't gonna help anything. What I didn't expect is how badly it hurt performance for both implementations.

*Q*: Why doesn't `Task.parSequence` use this same optimization?
*A*: `Task.parSequence` (as well as `Task.parSequenceN`) weren't designed to run on a list of tasks as huge as what we are providing in the test cases above. So it's not too unreasonable that it doesn't consider applying those optimizations. It does have some other pretty cool optimizations in it though.

*Q*: Okay, are you saying I should never use `Task.parSequence`?
*A*: Nah, `Task.parSequence` is still super legit. That being said, if you have more than 100,000 tasks, you might wanna use something with a little less overhead.

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala [hunk 1]
```diff
@@ -1,48 +1,75 @@
...
+    // create a stream of "observable events" matching the desired iterations
+    Observable.range(0, iterations)
+      // for each iteration, create a projection and check it
+      .mapParallelUnordered(parallelism)(_ => detectCircularity)
+      // sum the results of each detection
+      .sumL
+      // finally, calculate pi from the aggregate sum
+      .map(sum => {
+        4.0 * sum / (iterations - 1)
+      })
```
Comment (SwiftEngineer) [inline src/main/g8/src/main/scala/$package$/DemoParallelism.scala:72]: So here's some fun: We can actually still make this code several orders of magnitude more performant. All we would have to do is batch iterations into a single Task. Like this:

```
    // create a stream of "observable events" matching the desired iterations
    Observable.range(0, iterations)
      // batch iterations into groups of 100 thousand or less
      .bufferTumbling(100000)
      // for each batch of iterations, create a projections for each one and check it
      .mapParallelUnordered(parallelism)(batchIterations => {
        val batchedComputations = batchIterations.map(_ => detectCircularity)
        Task.sequence(batchedComputations).map(_.sum)
      })
      // collect all the results, then sum the results of each event
      .sumL
      // finally, calculate pi from the aggregate sum
      .map(sum => {
        4.0 * sum / (iterations - 1)
      })
```

**Q**: What kind of an effect does this have? 
**A**: Well, for starters, we can start running approximations with iterations of 10,000,000 in less than a second which is pretty legit. That's over 20x more performant than even the best case scenario of the NEW code, let alone the old.

**Q**: So why don't we do it?
**A**: If the goal of this project was to make an insanely performant, accurate, and long-running approximation of pi, we probably wouldn't be using graal native, since the JVM is gonna blow the doors off graal native in any application with a long runtime. I think the implementation of this PR does a couple things really well.
Firstly, it illustrates a clean way of performing parallel processing that works very well in graal native. Most importantly though, it has very reasonable and predictable performance based on the inputs you provide it. 

**TL;DR** If your goal is to show people a "real-world" example how parallelism in graal can be implemented and how well it performs, I think you'd have a hard time coming up with a better example than what we have here.

File: src/main/g8/build.sbt [hunk 2]
```diff
@@ -14,10 +14,33 @@ lazy val root = (project in file("."))
...
+      System.getProperty("os.name").toLowerCase match {
+        case mac if mac.contains("mac") =>
+          Nil
+        case win if win.contains("win") =>
+          Nil
+        case _ =>
+          // Forces statically-linked binary, compatible with libc (linux)
+          Seq("--static")
+      }
+    },
```
Comment (SwiftEngineer) [inline src/main/g8/build.sbt:35]: This is definitely one of the more dubious parts of this PR. In order to make it so folks don't use this template and then IMMEDIATELY run into the issue of trying to build the image with the `--static` flag on their native macbook pro laptop, I added this code to do a little check. It's a bad check, for reasons explained in the comments, but it does make this thing a lot more "plug and play" for folks who are less inclined to RTFM.

File: src/main/g8/build.sbt [hunk 3]
```diff
@@ -14,10 +14,33 @@ lazy val root = (project in file("."))
...
+
+    // Contains substitutions that are necessary for native image creation. These substitutions are required since monix
+    // depends on JCTools which contains a lot of code that can't be compiled by graal native:
+    libraryDependencies ++= Seq(
+      "science.doing" % "jctools-graal-native_20_2_0" % "1.0.0",
+      // Important: Keep the version of the "org.graalvm.nativeimage:svm" dependency synced with whatever version of
+      // graal we are using!
+      "org.graalvm.nativeimage" % "svm" % "20.2.0" % Provided
```
Comment (SwiftEngineer) [inline src/main/g8/build.sbt:43]: Shameless integration of my own lib into this PR :rofl:

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala [hunk 4]
```diff
@@ -1,48 +1,75 @@
...
+    _ <- Task {
+      logger.info(f"Will approximate Pi with \$iterations%,d iterations of random bubble-filling and up to \$parallelism%,d parallel processes...")
+    }
 
-    // Scheduler for parallelism:
-    implicit val s: Scheduler = Scheduler.forkJoin(parallelism, maxThreads = 1000)
+    // create scheduler resource based on desired parallelism (it will be closed once we have finished our calculations)
+    schedulerResource = Resource.make(acquire = Task {
+      Scheduler.forkJoin(parallelism = parallelism, maxThreads = 1000)
+    })(release = s => Task { s.shutdown() })
```
Comment (SwiftEngineer) [inline src/main/g8/src/main/scala/$package$/DemoParallelism.scala:29]: This thing would still exit just fine without shutting down this scheduler, but I like the fact that this function is now guaranteed to clean up after itself. That's nice if someone ever decides to copy pasta it into another project.

I'm always happy to advertise how awesome `cats.effect.Resource` is, but I concede that it is a bit off-topic for this project. For that reason, I'm on the fence about whether it's worth keeping around.

File: src/main/g8/src/main/scala/$package$/Main.scala [hunk 5]
```diff
@@ -12,16 +16,20 @@ import $package$.args.Command.{Hi, Pi}
   * @author sbchapin
   * @since 11/11/19.
   */
-object Main {
+object Main extends TaskApp with LazyLogging {
```
Comment (SwiftEngineer) [inline src/main/g8/src/main/scala/$package$/Main.scala:19]: I like me some TaskApp. The fact that it eliminates the confusion that can often come from combining eagerly-evaluated instances of Future with true monads like Task and that it requires returning an ExitCode are cool and all... but honestly I think the best part is how much more meaningful it makes cats.effect instances like Resource, Ref, MVar, etc.

File: src/main/g8/build.sbt (hunk 2 above)
Comment (sbchapin) [reply src/main/g8/build.sbt:35]: It's a better off-the-shelf experience for the average dev for sure, but I dislike this - given the same reasoning you provide in the comments (OS != Libc)

Another big weak spot is that `System.getProperty` may return null, so chaining it is scary.

File: src/main/g8/build.sbt (hunk 3 above)
Comment (sbchapin) [reply src/main/g8/build.sbt:43]: If we're gonna do this, I would suggest we either...
1. Put both of these projects under some other unified Github Org, or...
2. Pull it into this project as a dependent subroject.

As this stands, I don't want to be pulling in seemingly random other libraries _(random from a user's perspective, I know your stuff is fine)_.

That being said, since I'm getting back to this more than a year later _(sorry 😬)_ - perhaps GraalVM or Monix already have a patch for us.  I'll give everything a bump and see how far forward we can go without this.

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala [hunk 6]
```diff
@@ -1,48 +1,75 @@
...
-    val pi = approximatePi(iterations).runToFuture
+    // then use the scheduler to...
+    calculate <- schedulerResource.use(scheduler => {
+      // calculate an approximation of pi
+      approximatePi(parallelism, iterations)
+        // measure how long it took to perform the calculation
+        .timed
+        // set a timeout of 30 seconds
+        .timeout(30.seconds)
+        // insure the process runs on the provided scheduler
```
Comment (sbchapin) [inline src/main/g8/src/main/scala/$package$/DemoParallelism.scala]:
```suggestion
        // ensure the process runs on the provided scheduler
```
Yeah, I'm that kind of guy.

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala [hunk 7]
```diff
@@ -1,48 +1,75 @@
...
+        .executeOn(scheduler)
+    })
 
-    logger.info(s"Pi is roughly \${Await.result(pi, atMost = 30.seconds)}")
+    // log the results
+    timeTook = calculate._1
+    pi = calculate._2
```
Comment (sbchapin) [inline src/main/g8/src/main/scala/$package$/DemoParallelism.scala]:
```suggestion
    (timeTook, pi) = calculate
```

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala (hunk 1 above)
Comment (sbchapin) [reply src/main/g8/src/main/scala/$package$/DemoParallelism.scala:72]: > Q: What kind of an effect does this have?
> A: Well, for starters, we can start running approximations with iterations of 10,000,000 in less than a second which is pretty legit. That's over 20x more performant than even the best case scenario of the NEW code, let alone the old.

This is awesome.  To be honest, this makes me want to start a benchmarking competition.  Maybe we can have a "pi eating race" in a future version of this repo that includes a JMH harness in SBT?  JMH is actually useful, _and we can have some fun while we're at it, right?_

> Q: So why don't we do it?
> A: If the goal of this project was to make an insanely performant, accurate, and long-running approximation of pi, we probably wouldn't be using graal native, since the JVM is gonna blow the doors off graal native in any application with a long runtime. I think the implementation of this PR does a couple things really well.
> Firstly, it illustrates a clean way of performing parallel processing that works very well in graal native. Most importantly though, it has very reasonable and predictable performance based on the inputs you provide it.

Agreed.  Your "old new" implementation is already great, with the advantage of being very readable.  This is important, since the Monte Carlo method of Pi approximation is already a bit "abstract".  But if I were to highlight places for improvement in the "new new" implementation in your comment...
1. An observable range and a [tumblingBuffer](https://monix.io/api/current/monix/reactive/Observable.html#bufferTumbling(count:Int):monix.reactive.Observable[Seq[A]]) are ordered and deterministic - and both of those things have a cost in terms of memory allocation.  If the number of iterations on each process is fixed, we would need no buffering and no observing fundamentally - we are only using those Monix constructs out of convenience. This goes all the way down to the random calculation encapsulating a tuple - you technically don't even need that data structure.  I also think that for the purposes of pure optimization, more batches isn't necessarily better - eliminating arbitrary boundaries and delegating the allocation of resources to the scheduler is better.
2. The parallelism in `mapParallelUnordered(parallelism)` is also used to configure the scheduler this runs in, effectively coupling concurrency and parallelism.  Depending on which JRE this is running _(or on Javascript)_ and how the "Threads" are handled _([for instance](https://monix.io/docs/current/execution/scheduler.html#execution-model), a single process that delegates to setTimeout)_, this may have some adverse consequences to speed.

File: src/main/g8/src/main/scala/$package$/Main.scala (hunk 5 above)
Comment (sbchapin) [reply src/main/g8/src/main/scala/$package$/Main.scala:19]: I'd be more eager to accept this conversion if we had TaskApp and the app executions more bridged together in this implementation.  I dislike that when we parse the Args, it just dumps out a unit task rather than dealing with error propagated by incorrectly invoking the args.

Granted, the previous implementation wasn't much better in this regard.

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala (hunk 4 above)
Comment (sbchapin) [reply src/main/g8/src/main/scala/$package$/DemoParallelism.scala:29]: Keep it!

I learned something reading this, and the point of this template is to establish good and sane practices, _(and then let people willingly delete those good things)_.

**Review by sbchapin — COMMENTED** [review]
Totally understand if you've moved on, expecting about as much input as I gave you - but regardless, here's my review.  There's some stuff I would keep, some stuff I would change, some suggestions, and some questions.

File: src/main/g8/build.sbt (hunk 2 above)
Comment (SwiftEngineer) [reply src/main/g8/build.sbt:35]: :+1: super fair. I think at one point early on I started to use these lines at my work, but threw it away after just a few weeks because I started to see folks treating it like voodoo magic.

TL;DR: I've come to believe that Graal native just plain ain't for folks who don't like RTFMing, so I totally agree with you that we shouldn't make gambles/sacrifices just to give them a 20% chance at things working out of the box (and 80% chance of frustrating failure with confusing error messages like the potential `NullPointerException` you mentioned)

File: src/main/g8/build.sbt (hunk 3 above)
Comment (SwiftEngineer) [reply src/main/g8/build.sbt:43]: Don't even START to apologize on this one. It was shameless. As for the "perhaps GraalVM and Monix" already have a patch for this, I can tell you with almost 99.99% confidence they don't yet.

Monix is still unfortunately still using `sun.misc.Unsafe` via their `JCTools` underbelly. I've been watching the GraalVM releases like a hawk waiting for an update to be able to auto-substitute it and have yet to see anything. That's not to say they haven't patched a HUGE number of issues of course, but yeah, no dice.

**HOWEVER**, this matters not! If there is one thing I've learned about Graal Native, it's that sharing substitutions via libraries is a heck of a lot better of an idea on paper than it is in practice for two reasons:
1. By design, it obfuscates what pieces of the code are being substituted, and as such, which makes it really hard to tell if/when a substitution is no longer needed. It also contributes even more to the "voodoo magic-ness" of a project.
2. Versioning substitutions is just a pain in the butt in general. It's FAR easier to just copy the damn files around. As crazy as that may sound.

That brings up an interesting question: **Do you wanna include substitutions in this project?** On one hand, it makes your project a little bit fugly. Naturally, folks are gonna see your substitution files and either go "WTF" or at the very least "I don't think I'm in Kansas anymore". On the other, it shows folks that substitutions are INDEED an actual thing.

**So moving forward, I got three ideas, all with their own pros/cons:**
A) Kill the dependency on my crappy lib, move the substitutions into this project, and call it a day.
B) Kill the dependency on my crappy lib, remove the substitutions, remove Observable (which is what actually uses JCTools in the first place).
C) Kill the dependency on my crappy lib, remove the substitutions, replace `monix` with `cats-effect` (here is where you say: "say whaaaaaat?") or even `ZIO` (for the record though: I've yet to use ZIO personally).

Alright, alright, lemme justify option C a bit more, cause it's a :fire: take:
1. `cats-effect 3.x.x` is super easy to use (unfortunately, can't say the same for 2.x), and has YET to let me down when it comes to packing it into a native image. 
2. It means that when folks EXPAND on this code and start using new features, there is a significantly reduced chance of them running into a problem that requires substitutions to solve. If you keep `monix` and/or the substitutions, there is still a very high chance of someone calling a different part of monix's library, and consequently allowing the image process to discover a whole new area of code to run into problems with.

**TL;DR**: Rock, meet hard place. I don't like ANY of the options I presented, maybe you can find a fourth? In any case, you're totally right about the fact that `jctools-graal-native_20_2_0` should get yeeted. Good freakin' call :100:

File: src/main/g8/src/main/scala/$package$/DemoParallelism.scala (hunk 1 above)
Comment (SwiftEngineer) [reply src/main/g8/src/main/scala/$package$/DemoParallelism.scala:72]: :+1: on the idea of a benchmark competition. That sounds like a good bit of fun! I actually only recently started messing with JMH and I gotta say, I totally agree that it can be a ton of fun AND hella useful

Comment (SwiftEngineer) [reply src/main/g8/src/main/scala/$package$/DemoParallelism.scala:72]: > I were to highlight places for improvement in the "new new" implementation in your comment...

Some 5-head points there :+1: and even more reason to go about the benchmarking contest. Based on your comments, I'm already seeing yet ANOTHER set of ways that performance could be improved... particularly for graal-native. If there is gonna be a contest, I'm keeping my cards close to my chest :joy:

File: src/main/g8/src/main/scala/$package$/Main.scala (hunk 5 above)
Comment (SwiftEngineer) [reply src/main/g8/src/main/scala/$package$/Main.scala:19]: That's a good point that I totally missed. I wonder if that's a good reason to throw something like [decline's CommandIOApp](https://ben.kirw.in/decline/effect.html#interpreting-our-command-line-interface) into this shiz?

I will say this though: If the point of this template is truly to ([as you say](https://github.com/sbchapin/quickstart-scala-sbt-native-image.g8/pull/4#discussion_r747315337)) "establish good and sane practices", I do believe keeping any usages of `unsafeRunSync` out of sight should be among your top priorities.

Reasoning: I found that it's a heck of a lot easier to tell teams (that are just learning functional programming for the first time) to never ever use `unsafeRunSync`, then it is to help settle the arguments that inevitably come from a team finding it and deciding for themselves how often they should use it. 

Maybe that's just my experience though?

File: src/main/g8/build.sbt (hunk 2 above)
Comment (sbchapin) [reply src/main/g8/build.sbt:35]: I agree with you.

...but I also want it both ways:  I want to use this as a starter template (and to work out of the box), but I also want Graal.

I personally love this little template and want to use it for "all the things" that I start scala-wise, but that's just not feasible with its current state if I want those things to be stable.  By this point, I know all the things I need to do to gut the Graal chunks out if I need stability.  I also generally end up making a multi-project build for a lot of the use-cases I have.  Maybe there's a chance here to have a package-only sub-project that bakes all the crazy Graal crap on the side?

**Example subprojects:**
```
./app/         <- The usual assembly, contains Main.scala
./app-native/  <- The Graal/SVM & docker packaging of ./app/ 
./app-js/      <- Even more insanity
./common/      <- "Shared" libs
```
...thoughts?

File: src/main/g8/build.sbt (hunk 3 above)
Comment (sbchapin) [reply src/main/g8/build.sbt:43]: > Don't even START to apologize on this one.

I ain't.  I'm apologizing for how damn long it has taken me to get to reviewing this dope PR.  That's unacceptable.  But I ain't sorry for not pulling in your janky Unsafe patches.  On that topic...

> So moving forward, I got three ideas, all with their own pros/cons:
> A) Kill the dependency on my crappy lib, move the substitutions into this project, and call it a day.
> B) Kill the dependency on my crappy lib, remove the substitutions, remove Observable (which is what actually uses JCTools in the first place).
> C) Kill the dependency on my crappy lib, remove the substitutions, replace monix with cats-effect (here is where you say: "say whaaaaaat?") or even ZIO (for the record though: I've yet to use ZIO personally).

- **A** maybe, only short-term.
- **B** sucks.  Live in fear of the Unsafe Observable terror inside Monix? Obvious plant. No.
- **C** compelling.  You've shaken a bag full of cats-effect and Graal longer than anyone else that still lives _(2 years?  or more?)_.  If you're willing to contribute it, I'll actually review it in a timely manner.

File: src/main/g8/src/main/scala/$package$/Main.scala (hunk 5 above)
Comment (sbchapin) [reply src/main/g8/src/main/scala/$package$/Main.scala:19]: in-decline

For real though, gonna give Decline a try.  I've had minor beef with any arg parsing framework in Scala to-date - maybe this one will be different.

---

> I will say this though: If the point of this template is truly to (as you say) "establish good and sane practices", I do believe keeping any usages of unsafeRunSync out of sight should be among your top priorities.

Agreed.  Let's.

> Reasoning: I found that it's a heck of a lot easier to tell teams (that are just learning functional programming for the first time) to never ever use unsafeRunSync, then it is to help settle the arguments that inevitably come from a team finding it and deciding for themselves how often they should use it.

Often, it's inevitable that someone will make the decision to use unsafeRunSync or its ilk, despite your best intentions.  And it's typically a mismatch between your perceptions of their scope and their actual scope (and their perceptions of that scope). And that's being generous in ruling out bad actors.

I say this because **my immediate reaction is that any sane person would not use unsafeRunSync** _(at least in code expected to be stable and consistent)_.  But I would be wrong.  People get put in tight spots, and they make a sane decision to do this despite what I think.  They do it for a good reason (or at least _good to them_), and a lot of the time, it turns out fine.  My schadenfreude is when it does not, but this isn't "good".  

The productive action is to **educate people on why they shouldn't use unsafeRunSync**, and let him live their life with that knowledge.  I'm thinking something akin to [wat](https://www.destroyallsoftware.com/talks/wat), to illustrate really just what can go wrong.

Comment (SwiftEngineer) [reply src/main/g8/src/main/scala/$package$/Main.scala:19]: A :fire: meme there lol. Also +1 for usage of the word "schadenfreude". TIL that's a thing.

Comment (SwiftEngineer) [reply src/main/g8/src/main/scala/$package$/Main.scala:19]: > it's typically a mismatch between your perceptions of their scope and their actual scope (and their perceptions of that scope)

I'm presuming by "scope" here you mean the scope of the problem they are trying to solve. In which case, I can definitely get onboard with that statement.

Comment (SwiftEngineer) [reply src/main/g8/src/main/scala/$package$/Main.scala:19]: > The productive action is to educate people on why they shouldn't use unsafeRunSync, and let him live their life with that knowledge. I'm thinking something akin to wat, to illustrate really just what can go wrong.

10/10, love that philosophy :100:

File: src/main/g8/build.sbt (hunk 3 above)
Comment (SwiftEngineer) [reply src/main/g8/build.sbt:43]: > I ain't. I'm apologizing for how damn long it has taken me to get to reviewing this dope PR.

:joy: :joy: :joy: All gucci, my dude! Life happens.

Comment (SwiftEngineer) [reply src/main/g8/build.sbt:43]: > C compelling. 

Here it is if you wanna see what it could look like: https://github.com/sbchapin/quickstart-scala-sbt-native-image.g8/pull/5

File: src/main/g8/build.sbt (hunk 2 above)
Comment (SwiftEngineer) [reply src/main/g8/build.sbt:35]: Interesting line of thought :thinking: 

Is the motivation for shoving all the things in a big ole' multi-project build to make maintaining the templates less difficult? Or is it that you actually DO wanna have your cake and 20 different kinds of ~plates to eat it on~ ways to run it too? Both? Neither?

>  I also generally end up making a multi-project build for a lot of the use-cases I have.

When you say that, I'm assuming you're not SPECIFICALLY referring to "use-cases" as runtimes. If I'm wrong though, then that answers a part of my question above for sure.

Comment (sbchapin) [reply src/main/g8/build.sbt:35]: NGL, the idea of having one big multi-project that I can selectively delete things is appealing from a maintenance perspective.  

But for the use-case thing, I'll give some examples:

_Use-case 1:_  **Lambda functions on AWS**
I like to deploy lots of little jars, each of which encapsulate some domain logic for an endpoint or a binding of some sort like SQS or S3, and then version them separately.  They get connected via a consistent common/lib subproject behind them.

_Use-case 2:_ **CLI tool**
I like to make a "common" or "lib" subproject that I package separately, then bring into the actual application.  This is mainly to keep my dependencies separate and my interfaces clear so that I can reuse stuff that I slap into common with little effort.

Comment (SwiftEngineer) [reply src/main/g8/build.sbt:35]: I'm thinking of someone linking this project to their friend who maybe just asked something along the lines of "I really want to play around with Graal Native, but I don't know where to start". As it stands now, this project is like aimed JUST broadly enough, yet also specifically enough at someone who wants to get started with GraalVM with sbt. 

If we start throwing other stuff in the mix, does that "dillute" the graaly parts too much? Four sub-projects sure doesn't sound too bad to me, and I DO really like the idea of dog-piling because I get the sense there is at least a solid chance of being able to pull it off without adding too much complexity. 

TL;DR: If you think that, after that multi-module update, this project can be something that folks can link as a sort of "first hit is free" kind of lure to get folks into Graal Native, then I'm all in for this change :+1: otherwise, well, I'll be nervous, but of course, I'm still in :laughing:

**Comment by sbchapin** [comment]
Given the introduction of #5, I'm opting to close this PR.  However, I expect we'll probably be referring back to commentary here in future PRs.

Thank you for introducing these changes, and your further iteration to get the ideas to fruition!
